### PR TITLE
Moved windows workflow to Kubernetes hosted runner

### PR DESCRIPTION
**Pull request by Eliasj42**
Moved the windows MSVC workflow to a kubernetes hosted windows runner.  Also moved a file location to Azure instead of GCP.

workflow running successfully: https://github.com/iree-org/iree/actions/runs/11827489914/job/32955730437

Adresses https://github.com/iree-org/iree/issues/18813: runs windows workflow through a runner self hosted through kubernetes on a 32 vcpu machine.  Still need to implement sccache.

Job currently takes about 1 hour, may need to upgrade to 64 vcpu machine, although implementation of sccache will also save time

ci-exactly: windows_x64_msvc

### CODE REVIEW

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:1]: Please link logs of this running in the PR description.

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 1]
```diff
@@ -22,19 +22,20 @@ concurrency:
...
     steps:
       - name: "Checking out repository"
-        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+        uses: actions/checkout@v4.1.7
         with:
           submodules: true
+      - name: "Clean up build dir"
+        run: |
+          if (Test-Path -Path "C:\mnt\azure\build-windows") {Remove-Item -Path "C:\mnt\azure\build-windows" -Recurse -Force}
+          mkdir "C:\mnt\azure\build-windows"
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Why is this needed? Are these runner persistent?

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:1]: Tag related issues on PRs and explain in the PR description what work is done as well as what remains.

https://github.com/iree-org/iree/issues/18813 here

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 2]
```diff
@@ -22,19 +22,20 @@ concurrency:
...
       - name: "Setting up Python"
-        uses: actions/setup-python@f677139bbe7f9c59b41e40162b753c062f5d49a3 # v5.1.0
+        uses: actions/setup-python@v4
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Revert - pin to a hash and don't downgrade.

Did 5.1.0 not work for some reason?

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 3]
```diff
@@ -45,8 +46,16 @@ jobs:
       - name: "Installing requirements"
         run: choco install ccache --yes
       - name: "Configuring MSVC"
-        uses: ilammy/msvc-dev-cmd@0b201ec74fa43914dc39ae48a89fd1d8cb592756 # v1.13.0
-      - name: "Building IREE"
-        run: ./build_tools/cmake/build_all.sh "${BUILD_DIR}"
+        uses: ilammy/msvc-dev-cmd@v1.13.0
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Pin to a hash not a tag.

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 4]
```diff
@@ -45,8 +46,16 @@ jobs:
...
+      - name: "Installing iree reqs"
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: spell IREE consistently

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 5]
```diff
@@ -45,8 +46,16 @@ jobs:
...
+        run: |
+          choco install cmake
+          choco install ninja
+          Import-Module $env:ChocolateyInstall\helpers\chocolateyProfile.psm1
+          refreshenv
+      - name: "Building Iree"
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: spell IREE consistently

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 6]
```diff
@@ -45,8 +46,16 @@ jobs:
...
+        run: bash ./build_tools/cmake/build_all.sh "/c/mnt/azure/build-windows"
       - name: "Testing IREE"
-        run: ./build_tools/cmake/ctest_all.sh "${BUILD_DIR}"
+        run: bash ./build_tools/cmake/ctest_all.sh "/c/mnt/azure/build-windows"
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Use the env var for the build dir here?

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 7]
```diff
@@ -45,8 +46,16 @@ jobs:
...
+      - name: "Clean up build dir"
+        run: Remove-Item -Path "C:\mnt\azure\build-windows" -Recurse -Force
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Why is this needed? Are these runner persistent?

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:1]: Again, when editing only one workflows, set `skip-ci: reason` or the other git trailers in the PR description to avoid wasting CI resources.

https://iree.dev/developers/general/contributing/#ci-behavior-manipulation

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: .github/workflows/ci_windows_x64_msvc.yml (hunk 2 above)
Comment (Eliasj42) [reply .github/workflows/ci_windows_x64_msvc.yml]: yeah 5.1.0 didnt work, i can retry

File: .github/workflows/ci_windows_x64_msvc.yml (hunk 7 above)
Comment (Eliasj42) [reply .github/workflows/ci_windows_x64_msvc.yml]: yes

Comment (Eliasj42) [reply .github/workflows/ci_windows_x64_msvc.yml]: Although, I have it run at the beginning because if the workflow fails, it doesn't reach the cleanup step. Do you know how to add this to the process that runs after the workflow is completed, regardless of if it succeeds or not?

Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: Why? I thought our Linux cluster had ephemeral runners? Wouldn't the Windows cluster follow the same convention? (All of this should go on the tracking issue, PR description, etc. - I'm reading between the lines here when it would be much easier to review the design if it was documented)

Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: > Do you know how to add this to the process that runs after the workflow is completed, regardless of if it succeeds or not?

In the workflow file there are options: https://stackoverflow.com/questions/58858429/how-to-run-a-github-actions-step-even-if-the-previous-step-fails-while-still-f

For this sort of self-hosted runner setup though, it's usually safer to configure at the runner level, like https://github.com/iree-org/iree/blob/main/build_tools/github_actions/runner/config/hooks/post_job.sh (maybe ARC has a similar hook you can connect to)

File: .github/workflows/ci_windows_x64_msvc.yml (hunk 6 above)
Comment (Eliasj42) [reply .github/workflows/ci_windows_x64_msvc.yml]: I was just hard coding it because the workflow goes between bash and powershell which have different notations for path.  I'll rewrite with 2 env variables

**Comment by ScottTodd** [comment]
Please also follow the policies at https://iree.dev/developers/general/contributing/

* Follow the branch naming conventions

And general github best practices

* Capitalize the first word in the PR title (and follow similar writing guidelines for the PR description)
* Avoid force pushing once code review has started

File: .github/workflows/ci_windows_x64_msvc.yml (hunk 6 above)
Comment (Eliasj42) [reply .github/workflows/ci_windows_x64_msvc.yml]: Yeah, the env variable is being defined in bash and isnt carrying over to powershell

Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: Okay, writing out the path twice is fine for now.

We still shouldn't need the workflow itself to specify a functional path. That should be configured at the runner level.

With enough of these workarounds moved up into the runner config, we should be able to have the entire workflow use either powershell or bash as needed. We aren't going to have just one workflow - all of our workflows (like the Windows portions of https://github.com/iree-org/iree/blob/main/.github/workflows/pkgci_build_packages.yml, and workflows in other projects) need a reliable runner image to use, without this sort of patching.

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 8]
```diff
@@ -22,19 +22,20 @@ concurrency:
...
         uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
         with:
           submodules: true
+      - name: "Clean up build dir"
+        run: |
+          if (Test-Path -Path "$Env:BUILD_DIR_POWERSHELL") {Remove-Item -Path "$Env:BUILD_DIR_POWERSHELL" -Recurse -Force}
+          mkdir "$Env:BUILD_DIR_POWERSHELL"
       - name: "Setting up Python"
-        uses: actions/setup-python@f677139bbe7f9c59b41e40162b753c062f5d49a3 # v5.1.0
+        uses: actions/setup-python@f677139bbe7f9c59b41e40162b753c062f5d49a3
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Keep the comment.

We'd like to use tagged versions if we _could_, but for security reasons we use hashes. Developers can freely update which commit a given tag points to, while spoofing a hash is muuuuuch harder. Picking a hash lets us know that we are using a consistent version, so if a developer pushes a malicious package, we won't immediately start using it.

The middle ground we pick is using a commit hash but putting the associated tag in a comment.

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml:1]: Thanks for linking https://github.com/iree-org/iree/actions/runs/11619486082/job/32359204800.

Here's what I'm looking for / seeing:

## Timing

* Total time: 57m
* Time for checkout: 2m30s (good - GitHub's large hosted runners were taking 8-10 minutes sometimes)
* Installing deps: around 1 minute total (fine, low priority to optimize, way better than the 20 minutes installing Visual Studio / MSVC before)
* Build: 52 minutes
* Test: 1 minute

## Other misc things

* CMake version 3.30.5: fine. Could match what we use on Linux (CMake 3.23.2: https://github.com/iree-org/base-docker-images/blob/f97cc2f710ae11184742a5f4e28b044c4c481ecd/dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile#L31-L35), but we'll want to upgrade to 3.29+ soon anyways: https://github.com/iree-org/iree/issues/18598
* May want to start bumping the setup-python action to install and test 3.12 or 3.13 (https://github.com/iree-org/iree/issues/18652)
* Build is using MSVC as expected (sometimes it defaults to another toolchain :P)
    ```
    -- The C compiler identification is MSVC 19.41.34123.0
    -- The CXX compiler identification is MSVC 19.41.34123.0
    ```
* A few compiler warnings but nothing new I think: https://github.com/iree-org/iree/actions/runs/11619486082/job/32359204800#step:9:6760
* ccache reports that 98% of calls are "cacheable", so that's a good indicator that having a remote cache will work as we'd like:
    ```
    + ccache --show-stats
    Cacheable calls:    5826 / 5887 (98.96%)
      Hits:                0 / 5826 ( 0.00%)
        Direct:            0
        Preprocessed:      0
      Misses:           5826 / 5826 (100.0%)
    Uncacheable calls:    61 / 5887 ( 1.04%)
    Local storage:
      Cache size (GiB):  0.0 /  5.0 ( 0.00%)
    Remote storage:
      Hits:                0 / 5826 ( 0.00%)
      Misses:           5826 / 5826 (100.0%)
    ```
* Full list of tests ran as we'd expect:
    ```
    100% tests passed, 0 tests failed out of 1502
    ```

**Review by ScottTodd — COMMENTED** [review]
Thanks! This is much better now. If we have capacity, a 1 hour build _might_ be okay to enable on presubmit. I'd like to draw the line at 20 minutes, but we do have some 20 minute build + 20 minute test workflows. Another 20 minutes total on top of that isn't _too_ crazy.

File: .github/workflows/ci_windows_x64_msvc.yml (hunk 7 above)
Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: @saienduri or @Eliasj42 can either of you comment on persistent vs ephemeral runner usage here? We can start with the "clean up build dir" workflow steps, but longer term I'd like to see either:

1. Ephemeral runners that scale up/down, with no need for explicit clean up steps (the entire disk will go away)
2. Persistent runners that use a pre-job or post-job script configured at the runner / Docker(?) level that hides this detail from workflows

Users should be able to work with these runners without needing to understand the details of file paths, storage mounts, etc. As we add more workflows using these runners, maintaining such steps across each workflow and project will be difficult.

Comment (saienduri) [reply .github/workflows/ci_windows_x64_msvc.yml]: @ScottTodd These are still ephemeral runners that scale up and down based on github activity. The runners that are spun up use the same storage mount that we clean up after every run due to some storage problems with the ones that the runners are coming up with. Because we are completely clearing the storage every time, this should be fine.

Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: Oh, the OS partition is ephemeral but the storage is persistent? Do multiple runners share the same storage disk then...? Can you elaborate on "some storage problems" somewhere (GitHub tracking issue would be best, but I'll settle for a PR review thread like here or one of our internal chats) so we can investigate?

Comment (saienduri) [reply .github/workflows/ci_windows_x64_msvc.yml]: This was initially for the nightly case, but yes, as we scale this up and move it to presubmit, we should be moving to an ephemeral storage mount which is pretty easy to do in kubernetes (in which case we remove the cleanup step as well). @Eliasj42 has more context on the issues with the storage problems, but I remember he tried an exhaustive list of things over a couple of days which is when we decided to pivot to a mount

Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: Okay, but we'll be aiming for presubmit (`pull_request`) or at least postsubmit (`push`). Nightly already works today and is free, so this would "just" be speeding up that build and taking on some extra billing costs if we stop after just this PR.

**Review by ScottTodd — APPROVED** [review]

File: .github/workflows/ci_windows_x64_msvc.yml (hunk 7 above)
Comment (ScottTodd) [reply .github/workflows/ci_windows_x64_msvc.yml]: Postsubmit with a 1 hour build on a single runner may be reasonable to start putting mileage on this.

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 9]
```diff
@@ -22,17 +22,24 @@ concurrency:
...
+    runs-on: arc-runner-set
     env:
-      BUILD_DIR: build-windows
+      BASE_BUILD_DIR_POWERSHELL: C:\mnt\azure\b
     steps:
       - name: "Checking out repository"
         uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
         with:
           submodules: true
+      - name: "Create build dir"
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:36]: marking request changes until you're done testing directory creation/cleanup

LMK when this is ready for another review

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: tests/e2e/stablehlo_models/mnist_train_test/mnist_train_test.py
Comment (ScottTodd) [inline tests/e2e/stablehlo_models/mnist_train_test/mnist_train_test.py:1]: Please revert the changes in this file and move them to another PR. This test was disabled last night in https://github.com/iree-org/iree/commit/11f0099a2ff15d9f5b9e13e898485a746bbed52a. Generally, try to keep PRs focused on a single area, and don't group bug fixes together with feature work.

**Review by ScottTodd — COMMENTED** [review]
Can you share the latest logs of this workflow running, and put them in the PR description?

File: tests/e2e/stablehlo_models/mnist_train_test/mnist_train_test.py
Comment (Eliasj42) [reply tests/e2e/stablehlo_models/mnist_train_test/mnist_train_test.py:1]: ok the current logs are slow because we lost out vcpu quota somehow, but I'll link them

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 10]
```diff
@@ -15,24 +15,34 @@ on:
...
   group: ${{ github.workflow }}-${{ github.event.number || github.sha }}
   cancel-in-progress: true
 
 jobs:
   windows_x64_msvc:
-    runs-on: windows-2022
-    defaults:
-      run:
-        shell: bash
+    runs-on: arc-runner-set
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Ideally we could make this workflow more agnostic to the specific runner, so it can run in forks. One code pattern for that is https://github.com/openxla/stablehlo/blob/b1c1115f070493845ad92bd4a4461f77a2aa628d/.github/workflows/buildAndTestBazel.yml#L39
```yml
    runs-on: ${{ github.repository == 'openxla/stablehlo' && 'ubuntu-22.04-64core' ||  'ubuntu-22.04'  }}
```

Something to aim for later. All the build directory setup that is specific to these Azure runners will make that difficult and will limit workflow debugging to only developers with write access to this project.

A middle ground is possible where the `runs-on` field has a branch _and_ the `Create build dir` step only runs `if ${{ github.repository == ` or `if ${{ runs-on == arc-runner-set`.

Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: Can we call this `azure-windows-scale`, to match the existing `azure-linux-scale`? The name `arc-runner-set` doesn't suggest anything about the runner platform, size, features, etc.

How much effort would it be to rename it?

File: .github/workflows/ci_windows_x64_msvc.yml [hunk 11]
```diff
@@ -15,24 +15,34 @@ on:
...
-      run:
-        shell: bash
+    runs-on: azure-windows-scale
     env:
-      BUILD_DIR: build-windows
+      BASE_BUILD_DIR_POWERSHELL: C:\mnt\azure\b
+      SCCACHE_AZURE_CONNECTION_STRING: "${{ secrets.AZURE_CCACHE_CONNECTION_STRING }}"
+      SCCACHE_AZURE_BLOB_CONTAINER: ccache-container
+      SCCACHE_CCACHE_ZSTD_LEVEL: 10
+      SCCACHE_AZURE_KEY_PREFIX: "ci_windows_x64_msvc"
```
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:30]: FYI looks like we'll need some extra setup to use sccache on Windows beyond just running the build_all.sh script.

Linux uses `source ./build_tools/cmake/setup_sccache.sh`. Can write a similar script for Windows or work something back into `build_all.sh`.

Fine to keep this here, but it does make it seem like caching is expected to work, when I think it won't. Something to look into as part of https://github.com/iree-org/iree/issues/18557

**Review by ScottTodd — APPROVED** [review]
